### PR TITLE
Add a top level reinstall command to fwupdmgr

### PR DESCRIPTION
**Pull request by superm1**
I'm not proud of how many times I've done:
1. browse fwupd.org
2. find device
3. `fwupdmgr install $URL --allow-reinstall`

Or `fwupdmgr downgrade` just to have something to easily reinstall.

I can't be alone, so here is a way to make it easier :yellow_heart: 

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: src/fu-util.c [hunk 1]
```diff
@@ -2021,6 +2029,64 @@ fu_util_downgrade (FuUtilPrivate *priv, gchar **values, GError **error)
...
+	if (dev == NULL)
+		return FALSE;
+
+	/* try to lookup/match release from client */
+	rels = fwupd_client_get_releases (priv->client, fwupd_device_get_id (dev),
+					  NULL, error);
+	if (rels == NULL)
+		return FALSE;
+	for (guint j = 0; j < rels->len; j++) {
+		FwupdRelease *rel2 = g_ptr_array_index (rels, j);
```
Comment (hughsie) [inline src/fu-util.c]: `rel_tmp` would be the convention i think.

File: src/fu-util.c [hunk 2]
```diff
@@ -2021,6 +2029,64 @@ fu_util_downgrade (FuUtilPrivate *priv, gchar **values, GError **error)
...
+		if (fu_common_vercmp (fwupd_release_get_version (rel2),
+		    fu_device_get_version (dev)) == 0) {
+			rel = rel2;
```
Comment (hughsie) [inline src/fu-util.c]: I'd feel a lot more warm-and-fuzzy if this was a g_object_ref() and the toplevel rel was g_autoptr.

File: src/fu-util.c [hunk 3]
```diff
@@ -2525,6 +2591,13 @@ main (int argc, char *argv[])
 		     /* TRANSLATORS: sets something in daemon.conf */
 		     _("Modifies a daemon configuration value."),
 		     fu_util_modify_config);
+	fu_util_cmd_array_add (cmd_array,
+		     "reinstall",
+		     "[DEVICE-ID]",
+		     /* TRANSLATORS: command description */
+		     _("Re-install current firmware on the device."),
```
Comment (hughsie) [inline src/fu-util.c]: Is't _reinstall_ one word? If it's not, ignore me.

File: src/fu-util.c [hunk 4]
```diff
@@ -2021,6 +2029,64 @@ fu_util_downgrade (FuUtilPrivate *priv, gchar **values, GError **error)
...
+	g_signal_connect (priv->client, "device-changed",
+			  G_CALLBACK (fu_util_update_device_changed_cb), priv);
+	priv->flags |= FWUPD_INSTALL_FLAG_ALLOW_REINSTALL;
+	if (!fu_util_update_device_with_release (priv, dev, rel, error))
+		return FALSE;
+	fu_util_display_current_message (priv);
+
+	/* send report if we're supposed to */
+	remote_id = fwupd_release_get_remote_id (rel);
+	if (!fu_util_maybe_send_reports (priv, remote_id, error))
```
Comment (hughsie) [inline src/fu-util.c]: Semi-related question. Do you think reports are useful from version `n` to version `n`? Currently on the LVFS we just consider the target version, but I'm wondering if we want to filter the reports to be only upgrades server and client side?

Comment (superm1) [reply src/fu-util.c]: Well I think if someone is reinstalling something who isn't me or you it is interesting - it's a tell that they're trying to fix a problem that they think a reinstall would fix.

File: src/fu-util.c (hunk 3 above)
Comment (superm1) [reply src/fu-util.c]: It is. I think you're right.

File: src/fu-util.c (hunk 4 above)
Comment (hughsie) [reply src/fu-util.c]: Good point, agreed.

File: src/fu-util.c (hunk 3 above)
Comment (superm1) [reply src/fu-util.c]: Oh - elsewhere in the code it's also "re-install" (--allow-reinstall help).  I'll fixup both.

Comment (hughsie) [reply src/fu-util.c]: Great, thanks.

**Pull request merged**